File: ts/main.test.ts
import { expect, test } from "bun:test"
import { readdirSync } from "node:fs"
import Load, { type Extensions } from "./main"
import { testFormat } from "./testformat"

const testsDir = "../confetti/tests/conformance"

type testCase = {
	Name: string
	Input?: string
	Output?: string
	Extensions: Extensions
}

async function getCases(): Promise<testCase[]> {
	const cases: testCase[] = []
	const dir = readdirSync(testsDir).sort()

	const addTest = async (c: testCase, n: string, v: string) => {
		// read file
		const data = await Bun.file(`${testsDir}/${n}.${v}`).text()

		const strdata = data.replaceAll("\r\n", "\n")

		if (v === "conf") c.Input = strdata
		else if (v === "pass" || v === "fail") c.Output = strdata
		else if (v.startsWith("ext_")) {
			if (!c.Extensions) c.Extensions = {}

			switch (v.slice(4)) {
				case "c_style_comments":
					c.Extensions.CStyleComments = strdata
					break
				case "expression_arguments":
					c.Extensions.ExpressionArguments = strdata
					break
				case "punctuator_arguments":
					c.Extensions.PunctuatorArguments = strdata
					break
				default:
					throw new Error(`unknown extension ${v}`)
			}
		} else throw new Error(`unknown file type ${v}`)
	}

	dirloop: for (const entry of dir) {
		const split = entry.split(".")
		const [n, v] = split
		if (!n || !v) continue

		// search for the test case with the same name
		for (const c of cases) {
			if (c.Name !== n) continue
			await addTest(c, n, v)
			continue dirloop
		}

		const c: testCase = {
			Name: n,
			Extensions: {},
		}
		await addTest(c, n, v)
		cases.push(c)
	}

	for (const c of cases) {
		if (c.Input === undefined)
			throw new Error(`Test case ${c.Name} is missing input`)
		if (c.Output === undefined)
			throw new Error(`Test case ${c.Name} is missing output`)
	}

	return cases
}

function runConformanceTest(c: testCase) {
	const rin = c.Input
	const rout = c.Output
	const exts = c.Extensions
	if (!rin || !rout) return

	let out = ""
	try {
		out = testFormat(Load(rin, exts), 0)
	} catch (e) {
		if (!(e instanceof Error)) throw e
		out = `error: ${e.message}\n`
	}

	if (out !== rout)
		console.log(
			`Test case ${c.Name} failed:`,
			"\nInput:\n",
			Buffer.from(rin),
			"\nExpected:\n",
			Buffer.from(rout),
			"\nGot:\n",
			Buffer.from(out)
		)
	expect(out).toEqual(rout)
}

test("conformance", async () => {
	const cases = await getCases()

	for (let i = 144; i < cases.length; i++) {
		const c = cases[i]
		if (!c) continue

		console.log(
			`Test case ${i + 1}\nconfetti/tests/conformance/${c.Name}.conf`
		)
		runConformanceTest(c)
	}
})

File: go/main_test.go
package confetti

import (
	"fmt"
	"os"
	"strings"
	"testing"
)

const testsDir = "../confetti/tests/conformance"

type testCase struct {
	Name          string
	Input, Output *string
	Extensions    Extensions
}

func getCases(t *testing.T) (cases []*testCase, err error) {
	dir, err := os.ReadDir(testsDir)
	if err != nil {
		t.Fatalf("Failed to read tests directory: %v", err)
	}

	addTest := func(c *testCase, n, v string) error {
		// read file
		data, err := os.ReadFile(testsDir + "/" + n + "." + v)
		if err != nil {
			return fmt.Errorf("failed to read file %s.%s: %w", n, v, err)
		}

		strdata := strings.ReplaceAll(string(data), "\r\n", "\n")

		switch {
		case v == "conf":
			c.Input = &strdata
		case v == "pass", v == "fail":
			c.Output = &strdata
		case strings.HasPrefix(v, "ext_"):
			if c.Extensions == nil {
				c.Extensions = make(Extensions, 1)
			}

			switch v[4:] {
			case "c_style_comments":
				c.Extensions[ExtCStyleComments] = strdata
			case "expression_arguments":
				c.Extensions[ExtExpressionArguments] = strdata
			case "punctuator_arguments":
				c.Extensions[ExtPunctuatorArguments] = strdata
			default:
				return fmt.Errorf("unknown extension %s", v[4:])
			}
		default:
			return fmt.Errorf("unknown file type %s", v)
		}

		return nil
	}

dirloop:
	for _, entry := range dir {
		split := strings.Split(entry.Name(), ".")
		n, v := split[0], split[1]

		// search for the test case with the same name
		for _, c := range cases {
			if c.Name != n {
				continue
			} else if err = addTest(c, n, v); err != nil {
				return
			}
			continue dirloop
		}

		c := &testCase{Name: n}
		if err = addTest(c, n, v); err != nil {
			return
		}
		cases = append(cases, c)
	}

	for _, c := range cases {
		if c.Input == nil {
			t.Fatalf("Test case %s is missing input", c.Name)
		} else if c.Output == nil {
			t.Fatalf("Test case %s is missing output", c.Name)
		}
	}

	return
}

func runConformanceTest(c *testCase, t *testing.T) {
	rin, rout, exts := *c.Input, *c.Output, c.Extensions

	var out string
	if p, err := Load(rin, exts); err != nil {
		out = err.Error() + "\n"
	} else {
		out = testFormat(p, 0)
	}

	if rout != out {
		t.Fatalf("Output mismatch\n-- Expected:\n%s\n-- Got:\n%s", rout, out)
	}
}

func TestConformance(t *testing.T) {
	cases, err := getCases(t)
	if err != nil {
		t.Fatalf("Failed to get test cases: %v", err)
	}

	for i, c := range cases {
		if i != 144 {
			continue
		}
		
		t.Logf("Test case %d\nconfetti/tests/conformance/%s.conf", i+1, c.Name)
		runConformanceTest(c, t)
	}
}

func runReformatTest(c *testCase, t *testing.T) {
	rin, exts := *c.Input, c.Extensions

	var out string

	ts, err := lex(rin, exts)
	if err != nil {
		t.Fatal(err)
	} else if out, err = testReformat(ts); err != nil {
		t.Fatal(err)
	} else if rin != out {
		t.Fatalf("Output mismatch\n-- Expected:\n%s\n-- Got:\n%s", rin, out)
	}
}

func TestReformat(t *testing.T) {
	cases, err := getCases(t)
	if err != nil {
		t.Fatalf("Failed to get test cases: %v", err)
	}

	for i, c := range cases {
		if strings.HasPrefix(*c.Output, "error:") {
			continue
		}
		t.Logf("Test case %d\nconfetti/tests/conformance/%s.conf", i+1, c.Name)
		runReformatTest(c, t)
	}
}
